src/payload.rs

```diff
@@ -420,42 +382,164 @@ where
+    pub fn readline(&mut self) -> ReadUntil<S> {
+        self.read_until(b"\n")
+    }
+
+    /// Put unprocessed data back to the buffer
+    pub fn unprocessed(&mut self, data: Bytes) {
+        self.len += data.len();
+        self.items.push_front(data);
+    }
+
+    /// Get remaining data from the buffer
+    pub fn remaining(&mut self) -> Bytes {
+        self.items
+            .iter_mut()
+            .fold(BytesMut::new(), |mut b, c| {
+                b.extend_from_slice(c);
+                b
+            }).freeze()
+    }
+}
+
+pub struct ReadAny<'a, S> {
+    inner: &'a mut PayloadBuffer<S>,
+}
+
+impl<'a, S: 'a> Stream for ReadAny<'a, S>
+where
+    S: Stream<Item = Bytes, Error = PayloadError>,
+{
+    type Item = Bytes;
+    type Error = PayloadError;
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if let Some(data) = self.inner.items.pop_front() {
+            self.inner.len -= data.len();
+            Ok(Async::Ready(Some(data)))
+        } else {
+            match self.inner.poll_stream()? {
+                Async::Ready(true) => self.poll(),
+                Async::Ready(false) => Ok(Async::Ready(None)),
+                Async::NotReady => Ok(Async::NotReady),
+            }
+        }
+    }
+}
+
+pub struct ReadExact<'a, S> {
+    inner: &'a mut PayloadBuffer<S>,
+    size: usize,
+}
+
+impl<'a, S: 'a> Stream for ReadExact<'a, S>
+where
+    S: Stream<Item = Bytes, Error = PayloadError>,
+{
+    type Item = Bytes;
+    type Error = PayloadError;
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if self.size <= self.inner.len {
+            self.inner.len -= self.size;
+            let mut chunk = self.inner.items.pop_front().unwrap();
+            if self.size < chunk.len() {
+                let buf = chunk.split_to(self.size);
+                self.inner.items.push_front(chunk);
+                Ok(Async::Ready(Some(buf)))
+            } else if self.size == chunk.len() {
+                Ok(Async::Ready(Some(chunk)))
+            } else {
+                let mut buf = BytesMut::with_capacity(self.size);
+                buf.extend_from_slice(&chunk);
+
+                while buf.len() < self.size {
+                    let mut chunk = self.inner.items.pop_front().unwrap();
+                    let rem = cmp::min(self.size - buf.len(), chunk.len());
+                    buf.extend_from_slice(&chunk.split_to(rem));
+                    if !chunk.is_empty() {
+                        self.inner.items.push_front(chunk);
+                    }
+                }
+                Ok(Async::Ready(Some(buf.freeze())))
+            }
+        } else {
+            match self.inner.poll_stream()? {
+                Async::Ready(true) => self.poll(),
+                Async::Ready(false) => Ok(Async::Ready(None)),
+                Async::NotReady => Ok(Async::NotReady),
+            }
+        }
+    }
+}
+
+pub struct Copy<'a, S> {
```

Comment:
Since it is the name of std trait I think it would be better to rename it something like `CopyBuffer`

Reply:
Oh you're right, I'll change that